build/lib/src/resource/resource.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// A [Resource] encapsulates the logic for creating and disposing of some
+/// expensive object which has a lifecycle.
+///
+/// Actual [Resource]s should be retrieved using `BuildStep#fetchResource`.
+///
+/// Build system implementations should be the only users that directly
+/// instantiate a [ResourceManager] since they can handle the lifecycle
+/// guarantees in a sane way.
+class Resource<T> {
+  /// Factory method which creates an instance of this resource.
+  final FutureOr<T> Function() _create;
+
+  /// Optional method which is given an existing instance that is ready to be
+  /// disposed.
+  final FutureOr Function(T instance) _userDispose;
```

Comment:
Stable SDK analyzer still chokes on this. Need a private typedef for the field and the old function syntax in any method signatures.

Reply:
done